### PR TITLE
BaseModel serializable dict

### LINKED ISSUE #1409: JSON-like serialization on .dict()
# Feature Request

pydantic version: 1.5
pydantic compiled: True
install path: C:\Work\repos\data-purge\.venv\Lib\site-packages\pydantic
python version: 3.8.1 (tags/v3.8.1:1b293b6, Dec 18 2019, 23:11:46) [MSC v.1916 64 bit (AMD64)]
                     platform: Windows-10-10.0.18362-SP0
optional deps. installed: []

I'm using a framework (lambda) that expects the lambdas to return a simple dict to be passed along to another step.  This is done through marshaling that I don't really have a lot of access to modifying.  I'm using pydantic models for all of my de/serialization, and it had been working great to just say something to the effect of `return result.dict()`

However, now that I've added a datetime field to the model, I get a failure on: `Unable to marshal response: Object of type datetime is not JSON serializable`

Now, I know this is a pretty standard python issue with anything Json related.  But given that pydantic is such a great library for managing all this stuff, and it's not uncommon to have _another_ library expect to receive dicts, not json strings, I think it would be nice if we could have some way for `.dict()` to return a serializable format by default.

Essentially, I want to still have my pydantic model define the fields as a datetime, but I want the `.dict` step to convert those into a string format, rather than the datetime objects themselves.  Is this at all possible/reasonable?

```py
In [1]: from pydantic import BaseModel

In [2]: from datetime import datetime

In [3]: class Thing(BaseModel):
   ...:     x: datetime
   ...: 
In [5]: a = Thing(x="2020-01-01T00:00:00")

In [6]: a.dict()
Out[6]: {'x': datetime.datetime(2020, 1, 1, 0, 0)}

# preferred output:
Out[6]: {"x": "2020-01-01T00:00:00"}
...
```

Is this already possible?  Could this be added in some kind of way?  This isn't the first time I've hit this scenario, and I can't find an elegant way around it.  If only python actually was reasonable about datetimes, I wouldn't expect pydantic or something else to solve it :/

### PR DESCRIPTION
**Pull request by wozniakty**
## Change Summary

Added serializable function to BaseModel to create an immediately serializable dict format with configurable type conversion.

## Related issue number

Closes #1409 

## Checklist

* [x] Unit tests for the changes exist
* [ ] Tests pass on CI and coverage remains at 100%
* [ ] Documentation reflects the changes where applicable
* [ ] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

**Comment by wozniakty** [comment]
Notes: There are 3 mypy errors, an isort error, and 1 failing test on master according to my local run 🤷 I have been working on fixes but unsure if they should be included in here.

**Comment by wozniakty** [comment]
@samuelcolvin It won't be hard to switch this from a standalone method to a param on `dict` if that's strongly preferred, but I discussed some reasons for preferring it separate in the issue.  In summary:

- The state machine of a BaseModel to Json is, essentially `BaseModel > Dict w/ types from model > Dict w/ serializable types > JSON string` and I think having 3 separate behaviors that represent stepping to those steps in the chain seems more appropriate than overriding the behavior of dict based on a kwarg
- This ensures no change to the expected behavior of a dict, which should still be used internally for most cases.  Serializable becomes the specific function to use if & only if you need a specific type conversion for some serialization tool you're interop with

**Comment by wozniakty** [comment]
Weird, tests and mypy pass here but not locally 🤷

**Comment by wozniakty** [comment]
Any advice on if I should do some additional benchmarking would be appreciated.
Also, will update docs and make a change.md once interface is agreed upon.

**Comment by wozniakty** [comment]
Since samuel is out, would love if @PrettyWood or others could give a preliminary review.  Not sure who else might be best to tag.

**Review by PrettyWood — COMMENTED** [review]
First and foremost thanks for contributing to *pydantic*!
It's still quite unclear to me whether or not a new method should be added instead of a new kwarg for `dict` and `json` methods (which is actually added anyway). I read your [comment](https://github.com/samuelcolvin/pydantic/issues/1409#issuecomment-687368620), which makes sense but I feel like it's still a `dict` representation so having just an extra `kwarg` also makes complete sense to me.
Personnally I would first add the `serialize` kwarg (or another name) to `dict` method, benchmark if there is no perf impact and check afterwards. After you can always keep a `serialize` method, which would be just a helper method! WDYT?

File: pydantic/main.py [hunk 1]
```diff
@@ -618,17 +655,22 @@ def _get_value(
...
         if isinstance(v, BaseModel):
             if to_dict:
-                v_dict = v.dict(
-                    by_alias=by_alias,
-                    exclude_unset=exclude_unset,
-                    exclude_defaults=exclude_defaults,
-                    include=include,
-                    exclude=exclude,
-                    exclude_none=exclude_none,
+                v_dict = dict(
```
[the remark this answers is not in the document]
Comment (wozniakty) [reply pydantic/main.py:663]: It wasn't so much a perf concern to me but what seemed like an odd sort of double recursion where `_iter` called `dict` but really only as a pass-through back into `_iter`.  

It also seems like a somewhat risky dependency.  _Today_ `dict` is rather simply a conversion of `_iter` to dict, but if in a future state there was some reason to do something else with `dict`, they might have to unravel the  circular dependency on `dict` from `_iter`.  Even if perhaps I would have updated `dict` to take the additional param, it feels off to me for `dict` to depend on `_iter`, and `_iter` to depend on `dict`.

I am happy to defer to the style of others here if it's preferred.

**Comment by wozniakty** [comment]
I'm still leaning towards preferring a separate method over a kwarg for dict, but I decided to at least put up the alternative implementation on another branch that we could merge into this one if the consensus is to use the kwarg strategy: https://github.com/samuelcolvin/pydantic/compare/master...wozniakty:serializable-dict-kwarg

**Comment by wozniakty** [comment]
Someone let me know if I should switch to a PR from that other branch.  Both seem acceptable to me, though I prefer this implementation for reasons stated above.

**Comment by samuelcolvin** [comment]
Thanks so much for contributing.

I don't agree with this approach, I still think we should use what I described in https://github.com/samuelcolvin/pydantic/issues/1409#issuecomment-650116116

The difficult bit of this is not in adding the argument to dict, it's in finding a way to reuse logic from `pydantic_encoder` with full flexibility but without damaging performance of `.json()` - I still think that needs more thought.

**Pull request closed without merging**